Legend

angelanicholas opened the pull request:
- adds new victory legend component & tests
- demos:
![screen shot 2017-01-18 at 2 52 06 pm](https://cloud.githubusercontent.com/assets/2624467/22086489/c18d0242-dd8d-11e6-8381-c6ca8ffb6b59.png)

janesh-travolta committed:
Add initial victory-legend

janesh-travolta committed:
updated victory-legend

janesh-travolta committed:
merged master

janesh-travolta committed:
minor improvements

boygirl committed:
Merge pull request #116 from janesh-travolta/victory-legend

Add initial victory-legend

angelanicholas committed:
Merge branch 'master' into legend

* master: (164 commits)
  short cut equality checks
  simplify area
  add sCU to VictoryLabel
  adjust domain for padding when calculating padding
  add sCU to all primitive components
  Version 11.0.1 - VictorySharedEvents bug
  update changelog
  fallback container
  removed .DS_Store from project
  Version 11.0.0 - support VictorySelectionContainer
  update changelog
  set active on data when triggering toolitps
  add selection spec
  support active prop on primitive children
  add active
  dont include parent with all eventKey helper
  allow childName 0
  correct child-parent events
  lint
  extract selection logic
  ...

angelanicholas committed:
lint and spelling

angelanicholas committed:
address PR comments and other refactoring

angelanicholas committed:
update demo

angelanicholas committed:
update tests

angelanicholas committed:
reorder component methods

angelanicholas committed:
move leftOffset calc to getLegendState

angelanicholas committed:
don't pass props around unnecessarily

angelanicholas committed:
fix standalone demo

angelanicholas committed:
move demo for consistency

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -0,0 +1,252 @@
+import VictoryContainer from "../victory-container/victory-container";
+import Point from "../victory-primitives/point";
+
+const defaultStyles = {
```
@angelanicholas Can this stuff go into a theme instead? Can you add a theme prop so that the legends work with themes?

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -0,0 +1,252 @@
+
+  constructor(props) {
+    super(props);
+    this.state = this.getLegendState(props);
```
Other components calculate values in render and then pass some kind of `calculatedProps` object down to the methods that need them.

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -0,0 +1,252 @@
+        symbol: PropTypes.object
+      })
+    ),
+    containerComponent: PropTypes.element,
```
It would be nice to have legend take a `colorScale` prop so that it can auto-color a legend to match some component that uses one of our color scales _i.e._ pies and stacked bars

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -0,0 +1,252 @@
+      })
+    ),
+    containerComponent: PropTypes.element,
+    dataComponent: PropTypes.element,
```
VictoryLegend should work with events, but we can save that for v2 if you want.

boygirl reviewed (commented):
@angelanicholas this is looking good. I think the changes I would like to see before merging are 1) getting rid of state, and 2) making it work with themes (and then adding appropriate legend themes). If you get around to adding a colorScale that's great too, but a nice to have. Events can wait until v2 if you like.

angelanicholas committed:
remove state in favor of calculated props in render

angelanicholas committed:
consolidate orientation checks by storing in calculated props

angelanicholas committed:
add theme prop and move default styles

angelanicholas committed:
fix tests

angelanicholas committed:
add colorScale prop

angelanicholas committed:
alphabetize props

angelanicholas commented:
@boygirl Sounds good! I haven't really figured out what kind of events we'd want in legend yet, will file a ticket for that after merging. Removed state and added colorScale and theme props though! Ready for re-review.

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -0,0 +1,256 @@
+  getCalculatedProps() { // eslint-disable-line max-statements
+    const { role } = this.constructor;
+    const { data, orientation, theme } = this.props;
+    let { height, width } = this.props;
```
You should be able to define width / height in the theme too.

boygirl replied:
And padding

angelanicholas replied:
fixed!

angelanicholas committed:
alphabetize default props, remove default padding prop

angelanicholas committed:
allow height width padding to be specified in theme

boygirl commented on src/victory-legend/victory-legend.js:
```diff
@@ -111,16 +111,18 @@ export default class VictoryLegend extends React.Component {
     const labelStyles = [];
     let leftOffset = 0;
 
+    padding = Helpers.getPadding({ padding: padding || theme.padding });
```
You'll need to remove padding from default props, and make this check respect padding = 0. Also, if you remove padding from default props, this should have a fallback right in the code like props.padding || theme.padding || 0; except with existence checking that respects zero.

angelanicholas replied:
Helpers.getPadding has a 0 fallback already if `padding || theme.padding` is undefined, and padding was removed from defaultProps in the previous commit

boygirl commented:
🚀 Thanks @angelanicholas
